controllers/nim_account_controller.go

```diff
@@ -0,0 +1,484 @@
+			makeApiKeyUnknownCondition(account.Generation, initialMsg),
+			makeConfigMapUnknownCondition(account.Generation, initialMsg),
+			makeTemplateUnknownCondition(account.Generation, initialMsg),
+			makePullSecretUnknownCondition(account.Generation, initialMsg),
+		},
+	}
+
+	defer func() {
+		r.updateStatus(ctx, req.NamespacedName, *targetStatus)
+	}()
+
+	// fetch api secret
+	secretNs := account.Spec.APIKeySecret.Namespace
+	if secretNs == "" {
+		secretNs = account.Namespace
+	}
+	apiKeySecret := &corev1.Secret{}
+	apiKeySecretSubject := types.NamespacedName{Name: account.Spec.APIKeySecret.Name, Namespace: secretNs}
+	if err := r.Client.Get(ctx, apiKeySecretSubject, apiKeySecret); err != nil {
+		var msg string
+		if k8serrors.IsNotFound(err) {
+			msg = "api key secret was deleted"
+			logger.Info(msg)
+		} else {
+			msg = "failed to fetch api key secret"
+			logger.V(1).Error(err, "failed to fetch api key secret")
+		}
+		meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, msg))
+		r.cleanupResources(ctx, account)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	foundApiKeySec := "found api key secret"
+	logger.V(1).Info(foundApiKeySec)
+	meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, foundApiKeySec))
+
+	apiKeyBytes, foundKey := apiKeySecret.Data["api_key"]
+	if !foundKey {
+		err := fmt.Errorf("secret %+v has no api_key data", apiKeySecretSubject)
+		msg := "failed to find api key data in secret"
+		logger.V(1).Error(err, msg)
+		meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, msg))
+		r.cleanupResources(ctx, account)
+		return ctrl.Result{}, err
+	}
+	apiKeyStr := string(apiKeyBytes)
+	gotApiKey := "got api key"
+	logger.V(1).Info(gotApiKey)
+	meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, gotApiKey))
+
+	// fetch available runtimes
+	availableRuntimes, runtimesErr := utils.GetAvailableNimRuntimes()
+	if runtimesErr != nil {
+		msg := "failed to fetch NIM available custom runtimes"
+		logger.V(1).Error(runtimesErr, msg)
+		meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, msg))
+		r.cleanupResources(ctx, account)
+		return ctrl.Result{}, runtimesErr
+	}
+	runtimesOk := "got custom runtimes"
+	logger.V(1).Info(runtimesOk)
+	meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, runtimesOk))
+
+	// validate api key
+	if err := utils.ValidateApiKey(apiKeyStr, availableRuntimes[0]); err != nil {
+		msg := "api key failed validation"
+		logger.Error(err, msg)
+		meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, msg))
+		meta.SetStatusCondition(&targetStatus.Conditions, makeApiKeyFailureCondition(account.Generation, msg))
+		r.cleanupResources(ctx, account)
```

Comment:
What is the motivation for the cleanup in case of errors?

I have the thought that this can break momentarily a working environment during a [controller resync](https://github.com/kubernetes-sigs/controller-runtime/blob/8e29c6387e83b25306d59b87c29424f3e59ca255/designs/cache_options.md#L123).

Reply:
We don't clean up for all errors, only for errors relating to the API key validation. The motivation is to remove the model info from the cluster in case of a failed validation. We can remove this if you prefer. I don't think it was a requirement. It may be something that we did in TP and may no longer be required here since we use status conditions.

@mpaulgreen, @xieshenzh, Any thoughts?

Reply:
@xieshenzh wrote the code in the cronob. So, this is the code that deleted the associated resources - https://github.com/opendatahub-io/odh-dashboard/pull/2959/files#diff-654ddbcfb120f3fa423302828fb6b0e72b2a680a025a7a09bc24ccfc4ea2abd9R103-R110. I am not sure why we want to keep the associated configmap and template if api key is invalid? There could be downstream impact too... @yzhao583 @@olavtar any issues if we dont remove the secret, cm and template? But @israel-hdez this is a very common pattern.

Reply:
> I am not sure why we want to keep the associated configmap and template if api key is invalid?

If the API Key is invalid, I'm OK with the clean-up. Although, my question is general to the clean-up rather than specific to this `if` block; i.e. there are other cleanups before this one. And, also, I'm thinking mainly about re-syncs.

AFAIK, the controller will do a re-sync (i.e. it will run a reconcile cycle) automatically after a 10-hour period even if nothing has changed in the cluster. As this reconcile is using external services, a re-sync may encounter temporary issues (e.g. API Key validation, or fetching runtimes returning 500 status code, or the request timing out). Such temporary errors will lead to a clean-up, despite everything is still valid.

Since the controller is returning an error, controller-framework would trigger another reconcile quickly which may re-create again all resources if it succeeds. The small window of the deleted resources is my concern. I haven't seen how these resources are used later.

So, if you don't see an issue with the missing resources for a ¿short? window of time, then I'm good with the clean-up.

Reply:
> If the API Key is invalid, I'm OK with the clean-up. Although, my question is general to the clean-up rather than specific to this if block; i.e. there are other cleanups before this one. And, also, I'm thinking mainly about re-syncs.

We clean up in four scenarios, all of which qualify as a failure to validate the API key:
1. We failed to fetch the Secret encapsulating the API key; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L149).
2. We got the Secret, but it doesn't have the required key; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L162).
3. We failed to fetch NIM's available runtimes; in this case, we don't have runtime info to validate the key with; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L176).
4.  The actual validation fail; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L189).

> AFAIK, the controller will do a re-sync (i.e. it will run a reconcile cycle) automatically after a 10-hour period even if nothing has changed in the cluster. As this reconcile is using external services, a re-sync may encounter temporary issues (e.g. API Key validation, or fetching runtimes returning 500 status code, or the request timing out). Such temporary errors will lead to a clean-up, despite everything is still valid.

Legitimate concern. Perhaps we should reconsider the cleanups.

> Since the controller is returning an error, controller-framework would trigger another reconcile quickly which may re-create again all resources if it succeeds. The small window of the deleted resources is my concern. I haven't seen how these resources are used later.

The controller only returns an error for scenarios 2-4 from the ones described above. If the Secret is not found, no error is returned; see [here](https://github.com/TomerFi/odh-model-controller/blob/nim-account-controller/controllers/nim_account_controller.go#L150).

> So, if you don't see an issue with the missing resources for a ¿short? window of time, then I'm good with the clean-up.

I think we can live with that, but I would be happy if the dashboard people could comment on this, too.
@olavtar @andrewballantyne